lib/Selenium/CanStartBinary.pm

```diff
@@ -0,0 +1,318 @@
+        $args{remote_server_addr} = '127.0.0.1';
+    }
+
+    return { %args };
+}
+
+sub _build_binary_mode {
+    my ($self) = @_;
+
+    my $executable = $self->binary;
+    return unless $executable;
+
+    my $port = $self->port;
+    return unless $port != 4444;
+    if ($self->isa('Selenium::Firefox')) {
+        setup_firefox_binary_env($port);
+    }
+    my $command = $self->_construct_command($executable, $port);
+
+    system($command);
+    my $success = wait_until { probe_port($port) } timeout => 10;
+    if ($success) {
+        return 1;
+    }
+    else {
+        die 'Unable to connect to the ' . $executable . ' binary on port ' . $port;
+    }
+}
+
+sub shutdown_binary {
+    my ($self) = @_;
+
+    # TODO: Allow user to keep browser open after test
+    $self->quit;
+
+    if ($self->has_binary_mode && $self->binary_mode) {
+        my $port = $self->port;
+        my $ua = $self->ua;
+
+        $ua->get('127.0.0.1:' . $port . '/wd/hub/shutdown');
+
+        # Close the additional command windows on windows
+        if (IS_WIN) {
+            # Blech, handle a race condition that kills the driver
+            # before it's finished cleaning up its sessions
+            sleep(1);
+            $self->shutdown_windows_binary;
+        }
+    }
+}
+
+sub shutdown_windows_binary {
+    my ($self) = @_;
+
+    # Firefox doesn't have a Driver/Session architecture - the only
+    # thing running is Firefox itself, so there's no other task to
+    # kill.
+    return if $self->isa('Selenium::Firefox');
+
+    my $kill = 'taskkill /FI "WINDOWTITLE eq ' . $self->window_title . '"';
+    system($kill);
+}
+
+before DEMOLISH => sub {
+    my ($self) = @_;
+    $self->shutdown_binary;
+};
+
+sub DEMOLISH { };
+
+sub _construct_command {
+    my ($self, $executable, $port) = @_;
+
+    # Handle spaces in executable path names
+    $executable = '"' . $executable . '"';
+
+    my %args;
+    if ($executable =~ /chromedriver(\.exe)?"$/i) {
+        %args = (
+            port => $port,
+            'url-base' => 'wd/hub'
+        );
+    }
+    elsif ($executable =~ /phantomjs(\.exe)?"$/i) {
+        %args = (
+            webdriver => '127.0.0.1:' . $port
+        );
+    }
+    elsif ($executable =~ /firefox(-bin|\.exe)"$/i) {
+        $executable .= ' -no-remote ';
+    }
+
+    my @args = map { '--' . $_ . '=' . $args{$_} } keys %args;
+
+    # Handle Windows vs Unix discrepancies for invoking shell commands
+    my ($prefix, $suffix) = ($self->_cmd_prefix, $self->_cmd_suffix);
+    return join(' ', ($prefix, $executable, @args, $suffix) );
+}
+
+sub _cmd_prefix {
+    my ($self) = @_;
+
+    if (IS_WIN) {
+        my $prefix = 'start "' . $self->window_title;
+
+        # Let's minimize the command windows for the drivers that have
+        # separate binaries - but let's not minimize the Firefox
+        # window itself.
+        if (! $self->isa('Selenium::Firefox')) {
+            $prefix .= '" /MIN ';
+        }
```

Comment:
todo: fix bug where non-firefox, windows browsers don't have the closing quote in their start command